packages/@aws-cdk/aws-iam/lib/policy-statement.ts

```diff
@@ -7,6 +7,51 @@ import { mergePrincipal } from './util';
  * Represents a statement in an IAM policy document.
  */
 export class PolicyStatement {
+  /**
+   * Creates a new PolicyStatement based on the object provided.
+   * This will accept an object created from the `.toJSON()` call
+   * @param obj the PolicyStatement in object form.
+   */
+  public static fromJson(obj: any) {
+    const ensureArrayOrUndefined = (field: any) => {
+      if (field === undefined) {
+        return undefined;
+      }
+      if (typeof (field) !== "string" && !Array.isArray(field)) {
+        throw new Error("Fields must be either a string or an array of strings");
+      }
+      if (Array.isArray(field) && !field.reduce((acc: boolean, f: any) => (acc && typeof (f) === "string"), true)) {
+        throw new Error("Fields must be either a string or an array of strings");
+      }
+      return Array.isArray(field) ? field : [field];
+    };
+
+    const statement = new PolicyStatement({
+      actions: ensureArrayOrUndefined(obj.Action),
+      resources: ensureArrayOrUndefined(obj.Resource),
+      conditions: obj.Condition,
+      effect: obj.Effect,
+      notActions: ensureArrayOrUndefined(obj.NotAction),
+      notResources: ensureArrayOrUndefined(obj.NotResource)
+    });
+
+    statement.sid = obj.Sid;
+
+    // Since the principals are a more complex object, not just a string or an array of strings,
+    // then just passing them through on the constructor doesn't work.
```

Comment:
I feel like I am missing something... Why can't we just pass them through?

We can modify `PolicyStatementProps` to support specifying `principal` and `notPrincipal` and just assign them there? no?

Reply:
yeah, sid could be passed on the PolicyStatement constructor but any additional changes seemed like it was breaking existing contract conventions. `principals` and `notPrincipals` are IPrincipal interfaces. It doesn't make sense to change that interface and I generally come from a .net-typed system so the idea of having Array<IPrincipal|SomethingElse> never struck me.

Reply:
I guess I'm not seeing a way of adding this into the constructor without having to cascade the changes through the code. Additionally, at some point I have to convert an Object into one of the main IPrincipal types anyway... It seemed cleaner to keep that logic explicitly in the .fromJson

Reply:
I’ll try to grab your branch and propose something over it through a PR later this week.

Reply:
Perhaps a new implementation of IPrincipal that would take the JSON representation as a constructor parameter.

Reply:
Yes maybe, but I am not sure it’s worth the effort given the issue is basically the internal representation inside PolicyDocument so I think there should be a way to just short circuit within the class. Let me give it a shot.